File: test_suite.py
import time, urllib.request, urllib.parse, sys
sys.path.append('python-pkg')
import clickhouse_sidecar

client = clickhouse_sidecar.get_client(data_dir='debug-db-tests')
port = urllib.parse.urlparse(client.uri).port

def run_query(query, method="POST"):
    req = urllib.request.Request(f"http://127.0.0.1:{port}/", data=query.encode(), method=method)
    with urllib.request.urlopen(req) as res: return res.read()

run_query("CREATE TABLE IF NOT EXISTS bench_dirty (id UInt32) ENGINE = MergeTree ORDER BY id")
run_query("TRUNCATE TABLE bench_dirty")

print("Inserting 500 rows...")
t0 = time.time()
for i in range(500):
    run_query(f"INSERT INTO bench_dirty VALUES ({i})")
t1 = time.time()
print(f"500 async inserts took {t1-t0:.3f}s")

print("Selecting count...")
t0 = time.time()
count = run_query("SELECT count() FROM bench_dirty").decode().strip()
t1 = time.time()
print(f"Select count after inserts took {t1-t0:.3f}s, returned {count}")
assert count == "500", f"Expected 500, got {count}"

print("Selecting count again (should be fast, no flush)...")
t0 = time.time()
run_query("SELECT count() FROM bench_dirty")
t1 = time.time()
print(f"Second select took {t1-t0:.3f}s")
assert t1-t0 < 0.2, "Second select should not flush!"
print("OK Python Proxy!")

File: test_suite.mjs
import { getClient } from './node-pkg/index.mjs';
const client = await getClient({ dataDir: 'debug-db-node-tests' });
await client.exec({ query: 'CREATE TABLE IF NOT EXISTS bench_dirty (id UInt32) ENGINE = Memory' });
await client.exec({ query: 'TRUNCATE TABLE bench_dirty' });

console.log("Inserting 500 rows...");
const t0 = Date.now();
for (let i = 0; i < 500; i++) {
    await client.insert({ table: 'bench_dirty', values: [{id: i}], format: 'JSONEachRow' });
}
console.log(`500 async inserts took ${Date.now() - t0}ms`);

console.log("Selecting count...");
let t1 = Date.now();
let res = await client.query({ query: 'SELECT count() as c FROM bench_dirty', format: 'JSONEachRow' });
let json = await res.json();
console.log(`Select count took ${Date.now() - t1}ms, returned ${json[0].c}`);
if (json[0].c != 500) throw new Error("Expected 500");

console.log("Selecting count again (should be fast, no flush)...");
t1 = Date.now();
await client.query({ query: 'SELECT count() as c FROM bench_dirty', format: 'JSONEachRow' });
let dt = Date.now() - t1;
console.log(`Second select took ${dt}ms`);
if (dt > 200) throw new Error("Second select should not flush! Took " + dt + "ms");
console.log("OK Node Proxy!");
process.exit(0);
